python/lsst/cp/pipe/defects.py

```diff
@@ -0,0 +1,645 @@
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+#
+
+__all__ = ['FindDefectsTask',
+           'FindDefectsTaskConfig', ]
+
+import numpy as np
+import matplotlib.pyplot as plt
+from matplotlib.backends.backend_pdf import PdfPages
+import os
+import warnings
+
+import lsst.pex.config as pexConfig
+import lsst.pipe.base as pipeBase
+import lsst.afw.image as afwImage
+import lsst.meas.algorithms as measAlg
+import lsst.afw.math as afwMath
+import lsst.afw.detection as afwDetection
+import lsst.log as lsstLog
+import lsst.afw.display as afwDisplay
+import lsst.afw.cameraGeom.utils as cgUtils
+
+from lsst.ip.isr import IsrTask
+from .utils import NonexistentDatasetTaskDataIdContainer, SingleVisitListTaskRunner, countMaskedPixels, \
+    validateIsrConfig
+
+
+class FindDefectsTaskConfig(pexConfig.Config):
+    """Config class for defect finding"""
+
+    isrForFlats = pexConfig.ConfigurableField(
+        target=IsrTask,
+        doc="""Task to perform instrumental signature removal""",
+    )
+    isrForDarks = pexConfig.ConfigurableField(
+        target=IsrTask,
+        doc="""Task to perform instrumental signature removal""",
+    )
+    isrMandatoryStepsFlats = pexConfig.ListField(
+        dtype=str,
+        doc="isr operations that must be performed for valid results when using flats. " +
+        "Raises if any of these are False",
+        default=['doAssembleCcd', 'doFringe']
+    )
+    isrMandatoryStepsDarks = pexConfig.ListField(
+        dtype=str,
+        doc="isr operations that must be performed for valid results when using darks. " +
+        "Raises if any of these are False",
+        default=['doAssembleCcd', 'doFringe']
+    )
+    isrForbiddenStepsFlats = pexConfig.ListField(
+        dtype=str,
+        doc="isr operations that must NOT be performed for valid results when using flats. " +
+        "Raises if any of these are True",
+        default=['doAddDistortionModel', 'doBrighterFatter', 'doUseOpticsTransmission',
+                 'doUseFilterTransmission', 'doUseSensorTransmission', 'doUseAtmosphereTransmission']
+    )
+    isrForbiddenStepsDarks = pexConfig.ListField(
+        dtype=str,
+        doc="isr operations that must NOT be performed for valid results when using darks. " +
+        "Raises if any of these are True",
+        default=['doAddDistortionModel', 'doBrighterFatter', 'doUseOpticsTransmission',
+                 'doUseFilterTransmission', 'doUseSensorTransmission', 'doUseAtmosphereTransmission']
+    )
+    isrDesirableSteps = pexConfig.ListField(
+        dtype=str,
+        doc="isr operations that it is advisable to perform, but are not mission-critical." +
+        " WARNs are logged for any of these found to be False.",
+        default=['doBias']
```

Comment:
This is a combination of things. I think it was partly because I did this once before and therefore cargo-culted it, but also, things like calib production are scenarios where you want it to be possible for the user to set _some_ config options, but not allow certain others that would render the output meaningless, _i.e._ to validate the options that have been chosen, rather than setting defaults and trusting people not to override them in a way that is Bad. Happy to discuss this properly OOB though, as I realise this is inelegant in the extreme.

Reply:
For now though, I'll mark this as resolved, and assume it's OK to keep unless I hear otherwise...